File: leetcode/0030-substring-with-concatenation-of-all-words/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.isEmpty() || words == null || words.length == 0) {
            return result;
        }

        int wordLen = words[0].length();
        int numWords = words.length;
        int windowLen = wordLen * numWords;

        if (windowLen > s.length()) {
            return result;
        }

        Map<String, Integer> wordCount = new HashMap<>();
        for (String w : words) {
            wordCount.merge(w, 1, Integer::sum);
        }

        for (int start = 0; start <= s.length() - windowLen; start++) {
            Map<String, Integer> seen = new HashMap<>();
            boolean ok = true;

            for (int i = 0; i < numWords; i++) {
                int chunkStart = start + i * wordLen;
                String chunk = s.substring(chunkStart, chunkStart + wordLen);

                if (!wordCount.containsKey(chunk)) {
                    ok = false;
                    break;
                }

                int count = seen.merge(chunk, 1, Integer::sum);
                if (count > wordCount.get(chunk)) {
                    ok = false;
                    break;
                }
            }

            if (ok) {
                result.add(start);
            }
        }

        return result;
    }
}

File: leetcode/0030-substring-with-concatenation-of-all-words/solution.py
from collections import Counter
from typing import List


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_len = len(words[0])
        num_words = len(words)
        window_len = word_len * num_words

        if window_len > len(s):
            return []

        word_count = Counter(words)
        result = []

        for start in range(0, len(s) - window_len + 1):
            seen = Counter()
            ok = True

            for i in range(num_words):
                chunk_start = start + i * word_len
                chunk = s[chunk_start:chunk_start + word_len]

                if chunk not in word_count:
                    ok = False
                    break

                seen[chunk] += 1
                if seen[chunk] > word_count[chunk]:
                    ok = False
                    break

            if ok:
                result.append(start)

        return result
